### DISCUSSION TITLE
X20S Port

### DISCUSSION THREAD

**Discussion by robthomson** [EdgeTX/edgetx#1725] [General]
Ok.. can I point you to this page:

https://openrcforums.com/forum/viewtopic.php?f=7&t=13679

Seems MikeB has managed to work out 99% of the X20/X20S hardware base.

Can we consider a port?

**Comment by brainbubblersbest** [comment]
Hello Rob.
As i wrote to you on FB:
MikeB has kind of a NDA with FRSKY.
Hes not allowed to Publish any Hardware Details FRSKY delivered to him.
I think @raphaelcoeffic would be open for that. But i think FRSKY doesnt want to share Technical Details with the other Radio Companys here. 

@[Dual Boot?](https://www.youtube.com/watch?v=661QG88XkHk)

Anyway... Dual Boot is awesome

**Comment by robthomson** [comment]
I imagine this falls into a grey area.. Believe Mike may have looked at the assembler code to work out what inferfaces to use on the chip!

**Comment by brainbubblersbest** [comment]
Maybe He is sort of a Ultra Brain, can read the Garbage output from a Decompiler and in addition He done the work to get this running in quite no Time.
There is still no Support from Frsky. Not for any Technical requests for Bug tracking. And no Details how to get the Tandem HF Hardware running.

I would love to see edgetx running on that Radio too. 😅👌

**Comment by rotorman** [comment]
X20/X20S uses STM32H7 microcontroller. Presently EdgeTX uses StdPeriphLibs for low-level interface and it does not support STM32H7 (StdPeriphLibs support ends with STM32F4). Without a change in EdgeTX to STM32 HAL/LL Libs (planned towards v2.8: https://raw.githubusercontent.com/EdgeTX/edgetx.github.io/master/images/roadmap.png ), X20/X20S support is pretty much not possible yet.
Now please do not try to read between the lines of the above text - this does NOT mean that X20/X20S support is planned in v2.8 or afterwards - as w/o cooperation from the radio manufacturer, there is a considerable effort required to get the code running on such a radio. And with numerous other radio manufacturers with beautiful radios and genuine interest of having EdgeTX on their radios and support towards the project, the time of developers is better invested towards supporting these radios than wasting time with a manufacturer that has turned it's back to OTX/ETX.

**Comment by donkeydog67** [comment]
Any update on EdgeTX support for X20, X20S, X20 HD series. I think someone know if it will be possible in the future or if technically it is possible to bring EdgeTX to X20 series. Many people is still waiting the answer of this issue.

**Reply by gagarinlg** [reply]
There are plans to support STM32F7 MCUs but adding support for a radio without support from the manufacturer creates a lot of costs and extra work. As of now, I think this is unlikely.

**Reply by donkeydog67** [reply]
> There are plans to support STM32F7 MCUs but adding support for a radio without support from the manufacturer creates a lot of costs and extra work. As of now, I think this is unlikely.

If STM32F7 MCUs compatibility is added, I think It will be easy for them to make it compatible with EdgeTX. The problem is that there are many people still using FrSky products. I hope developers find a way to make it compatible with EdgeTX.

**Reply by pfeerick** [reply]
While some of the core technical barriers may no longer be an issue in a future release, the main blocking reason of no support from Frsky still remains. And given they have moved away from OpenTX and are pushing their own Ethos OS I don't see that situation changing any time soon. So I personally have no time for Frsky for new hardware support until their business practices change. They used to be good, but I think they got too big/successful. If/when we get to a point their hardware *can* be supported, it will most likely be up to users of their products to tell them what is wanted ;)

**Reply by gagarinlg** [reply]
Same here

**Comment by brainbubblersbest** [comment]
Best Support someone can supply would be two Radios. Or one Radio plus the Electronic in spare parts. Then the internals can be re engineered like risto does it with the tx16s. And the Radio can be used for developing and debugging. So if You would like to Support this. Spend some Hardware or collect the Money for this, is the bare minimum someone can do without coding. I dont think one dev here wold spend the money for this by motivation. And thats pure logic.
Frsky collaborate with ersky.
But dont even talk to edgetx.

**Reply by pfeerick** [reply]
The biggest hurdle IMO is finding that person with both the skills and interest in doing the RE to enable hardware support for that hardware... the access to hardware / financial aspect is secondary as that could be organised through the edgetx opencollective as a project. But that is naturally only viable once the ETX codebase can even support that hardware. ;)

**Reply by gagarinlg** [reply]
Don't forget the time to do it. Currently we have more thing we want or have to do than available time for the hobby.

**Reply by donkeydog67** [reply]
> Don't forget the time to do it. Currently we have more thing we want or have to do than available time for the hobby.

When EdgeTX codebase start supporting that hardware probably in version 2.9, I will come back and reopen the issue.

**Reply by pfeerick** [reply]
Atm it is looking like it will be later than 2.9 before that happens - but yes, definitely revisit this again then to see if anything has changed. :)

**Reply by brainbubblersbest** [reply]
> The biggest hurdle IMO is finding that person with both the skills and interest in doing the RE to enable hardware support for that hardware... the access to hardware / financial aspect is secondary as that could be organised through the edgetx opencollective as a project. But that is naturally only viable once the ETX codebase can even support that hardware. ;)

Risto created an excellent howto (as always) how to delaminate a pcb and create Digital Information to work with. Maybe someday, someone does that just because Frsky dont want to see its new Topline Radio get "hacked".
For fun. 😂

**Comment by raphaelcoeffic** [comment]
Let me summarise the situation regarding new FrSky radios:
- EdgeTX welcomes vendor supported code contributions
- EdgeTX welcomes community supported contributions
- EdgeTX core team cannot and will not provide support for FrSky radios which are not yet supported as long as the necessary information, hardware and code contributions from either community or vendor are not provided.

**Reply by pfeerick** [reply]
And while this comment is focused on FrSky given this is in relation to a FrSky discussion, IMO you can substitute _any_ other manufacturer or radio and the same applies. We only have so many spare hours, so will naturally focus on the hardware we ourselves have, or have support from the manufacturers/vendors.

**Comment by rwijnhov** [comment]
Will this change things?
https://www.rcgroups.com/forums/showpost.php?p=50957367&postcount=228

**Reply by gagarinlg** [reply]
Not necessarily. We will see.

**Comment by Crunch69** [comment]
Please do a X20S port I love and support $$ open source!!   I currently use ETX and OTX on several radios would like to add new protocols and the only way to do that is with new hardware

**Reply by gagarinlg** [reply]
There will be STM32H7 based radios for EdgeTX. We will not add support for newer FrSky radios. They do it want this and there are load of alternatives to their radios. We do not want to spend our time in reverse engineering FrSky HW.
